### PR TITLE
🏛 Async State foundations + migrate useFirstTransactionTimestamp 

### PR DESCRIPTION
**Pull request by jxom**
Fixes RNBW-4402

### What changed (plus any additional context for devs)

This PR sets up our foundations for managing Async State in Rainbow ([5642ae7](https://github.com/rainbow-me/rainbow/pull/4164/commits/5642ae724bf0b5b4071aba0057b42b51b7320b92) & [0bcf21e](https://github.com/rainbow-me/rainbow/pull/4164/commits/0bcf21e191160bd04244134b40b418824d7bf7e0)), along with migrating an async Hook to demonstrate the usage of the foundation ([ca89d26](https://github.com/rainbow-me/rainbow/pull/4164/commits/ca89d26f343e6c5aee63b9a2ffc6fdbc7218e06c)).

I have also added a couple of template files that consumers can copy + pasta, then find + replace to create their own queries/mutations.

### CODE REVIEW

File: src/react-query/PersistQueryClientProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
+// https://github.com/TanStack/query/blob/main/packages/react-query-persist-client/src/PersistQueryClientProvider.tsx
```
Comment (jxom) [inline src/react-query/PersistQueryClientProvider.tsx:1]: Had issues where React Query was importing a different instance of the QueryClientProvider causing children to not consume the correct queryClient. I had a lot of trouble trying to figure out why, so I just copied the source.

Comment (estrattonbailey) [reply src/react-query/PersistQueryClientProvider.tsx:1]: Do you happen to have an example of the errors you were getting? Seems odd.

But yeah if we need to go this route, mind just updating this comment to reference the permalink to this file? That way we'll get the snapshot of what the file looked like when you copied it.

File: src/react-query/createQueryKey.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  /** Configuration for the query key. */
+  config: {
+    /**
+     * A storage version number for the query.
+     * If a query is stored against a storageVersion,
+     * which is later changed, the cache will bust for this query,
+     * and it will be invalidated.
+     */
+    storageVersion?: number;
```
Comment (estrattonbailey) [inline src/react-query/createQueryKey.ts]: Nice!

File: src/react-query/queryClient.ts [hunk 3]
```diff
@@ -1,7 +1,31 @@
 import { QueryClient } from '@tanstack/react-query';
+import AsyncStorage from '@react-native-async-storage/async-storage';
+import { createAsyncStoragePersister } from '@tanstack/query-async-storage-persister';
+import { PersistQueryClientOptions } from '@tanstack/react-query-persist-client';
 
 export const queryClient = new QueryClient({
   defaultOptions: {
-    queries: {},
+    queries: {
+      cacheTime: Infinity,
```
Comment (estrattonbailey) [inline src/react-query/queryClient.ts:9]: Yessss. For others' reference: https://www.notion.so/rainbowdotme/Async-State-RFC-711800896f9a458889fc4ffad4de271a#41a61b0c4ed24281bea7f8360c97f56a

File: src/react-query/queryClient.ts [hunk 4]
```diff
@@ -1,7 +1,31 @@
...
+
+export const persistOptions: Omit<PersistQueryClientOptions, 'queryClient'> = {
+  persister: asyncStoragePersister,
+  dehydrateOptions: {
+    shouldDehydrateQuery: query =>
+      Boolean(
+        // We want to persist queries that have a `cacheTime` of above zero.
+        query.cacheTime !== 0 &&
+          // We want to persist queries that have `storageVersion` in their query key.
+          (query.queryKey[2] as { storageVersion?: number })?.storageVersion
```
Comment (estrattonbailey) [inline src/react-query/queryClient.ts]: Thank you for the comments 🙏

File: src/resources/_template/_templateMutation.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * This template follows the patterns described in the Async State RFC:
+ * https://www.notion.so/rainbowdotme/Async-State-RFC-711800896f9a458889fc4ffad4de271a
+ *
+ * How to use this template:
+ *
+ * Step 1. Find & replace "TemplateResource" with your PascalCase query name (e.g. SignMessage)
+ * Step 2. Find & replace "templateResource" with your camelCase query name (e.g. signMessage)
+ * Step 3. Define your mutation arguments under "// Mutation Types"
+ * Step 4. Define your mutation function under "// Mutation Function"
+ */
```
Comment (estrattonbailey) [inline src/resources/_template/_templateMutation.ts:11]: 😍

File: src/resources/transactions/firstTransactionTimestampQuery.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  );
+
+export type FirstTransactionTimestampQueryKey = ReturnType<
+  typeof firstTransactionTimestampQueryKey
+>;
+
+// ///////////////////////////////////////////////
+// Query Function
+
+const queue = new PQueue({ interval: 1000, intervalCap: 5 });
```
Comment (estrattonbailey) [inline src/resources/transactions/firstTransactionTimestampQuery.ts:40]: Unrelated to this PR: I wonder if we'll still need this to rate limit these requests 🤔

File: src/resources/transactions/firstTransactionTimestampQuery.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+export async function firstTransactionTimestampQueryFunction({
+  queryKey: [{ addressOrName }],
+}: QueryFunctionArgs<typeof firstTransactionTimestampQueryKey>) {
+  if (!addressOrName) return null;
```
Comment (estrattonbailey) [inline src/resources/transactions/firstTransactionTimestampQuery.ts:45]: Nit: is there a reason to return `null` over `undefined`? Sometimes plays nicer with TS I think, but not a huge deal.

File: src/components/ens-profile/RecordTags/RecordTags.tsx [hunk 8]
```diff
@@ -36,7 +36,7 @@ export default function RecordTags({
   records,
   show,
 }: {
-  firstTransactionTimestamp?: number;
+  firstTransactionTimestamp?: number | null;
```
Comment (estrattonbailey) [inline src/components/ens-profile/RecordTags/RecordTags.tsx:39]: Like maybe if we just return `undefined` we can simplify stuff like this

File: src/react-query/createQueryKey.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  config: {
+    /**
+     * A persister version number for the query.
+     * If a persisterVersion exists, this means that this query
+     * will be stored in AsyncStorage.
+     * When a query is stored against a persisterVersion,
+     * and is later changed, the cache will bust for this query,
+     * and it will be invalidated.
+     */
+    persisterVersion?: number;
```
Comment (estrattonbailey) [inline src/react-query/createQueryKey.ts:21]: Yesss it's perfect

**Review by estrattonbailey — APPROVED** [review]
Incredible work!

File: src/resources/transactions/firstTransactionTimestampQuery.ts (hunk 6 above)
Comment (jxom) [reply src/resources/transactions/firstTransactionTimestampQuery.ts:40]: Good question! I don't think so.

File: src/resources/transactions/firstTransactionTimestampQuery.ts (hunk 7 above)
Comment (jxom) [reply src/resources/transactions/firstTransactionTimestampQuery.ts:45]: Unfortunately, React Query doesn't support returning `undefined` as it is not serializable :(

Comment (estrattonbailey) [reply src/resources/transactions/firstTransactionTimestampQuery.ts:45]: Works for me!

**Comment by jxom** [comment]
> Do you happen to have an example of the errors you were getting? Seems odd.

No errors in particular. But the children of this provider were not able to get the correct reference of the query client instance. So it seems that there may be duplicate instances of the query client being instantiated.

File: src/resources/_template/_templateQuery.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+ *
+ * Step 1. Find & replace "TemplateResource" with your PascalCase query name (e.g. EnsAvatar)
+ * Step 2. Find & replace "templateResource" with your camelCase query name (e.g. ensAvatar)
+ * Step 3. Define your query arguments under "// Query Types"
+ * Step 4. Define your query key under "// Query Key"
+ * Step 5. Define your query function under "// Query Function"
+ * Step 6. Define your query hook under "// Query Hook"
+ * Step 7. (Optional) Define your query prefetcher under "// Query Prefetcher"
+ * Step 8. (Optional) Define your query fetcher under "// Query Fetcher"
+ * Step 9. Delete these steps! :D
```
Comment (estebanmino) [inline src/resources/_template/_templateQuery.ts:15]: 👨‍🍳

**Review by estebanmino — APPROVED** [review]
🔥 🔥 🔥 
i also tried android and ios just in case, working as expected

**Pull request merged**